imap_processing/cdf/tests/test_cdf_attribute_manager.py

```diff
@@ -1,58 +1,322 @@
+    )
+    with pytest.raises(KeyError):
+        assert cdf_manager.global_attributes["imap_test_T1_test"]["Project"]
+
+    # Trying to update a default global using get_global_attributes does not work.
+    # For example, thing about DOI event.
+
+    # Testing that required schema keys are in get_global_attributes
+    for attr_name in cdf_manager.global_attribute_schema.keys():
+        required_schema = cdf_manager.global_attribute_schema[attr_name]["required"]
+        if required_schema is True:
+            assert attr_name in test_get_global_attrs.keys()
+
+
+def test_add_global_attribute():
+    # Initialize CdfAttributeManager object which loads in default info
+    cdf_manager = CdfAttributeManager(Path(__file__).parent.parent / "config")
+
+    # Change filepath to load test global attributes
+    cdf_manager.source_dir = Path(__file__).parent.parent / "tests"
+    cdf_manager.load_global_attributes("imap_default_global_test_cdf_attrs.yaml")
+    cdf_manager.load_global_attributes("imap_test_global.yaml")
+
+    # Changing a dynamic global variable
+    cdf_manager.add_global_attribute("Project", "Test Project")
+    assert cdf_manager.global_attributes["Project"] == "Test Project"
+    test_get_global_attrs = cdf_manager.get_global_attributes("imap_test_T1_test")
+    assert test_get_global_attrs["Project"] == "Test Project"
 
 
 def test_variable_attribute():
+    """
+    Test function that covers:
+        load_variable_attributes
+        get_variable_attributes
+    """
+
+    # Creating CdfAttributeManager object, loading in default data
     cdf_manager = CdfAttributeManager(Path(__file__).parent.parent / "config")
-    cdf_manager.load_global_attributes("imap_default_global_cdf_attrs.yaml")
-    cdf_manager.load_variable_attributes("imap_mag_l1a_variable_attrs.yaml")
+    cdf_manager.source_dir = Path(__file__).parent.parent / "tests"
+    cdf_manager.load_global_attributes("imap_default_global_test_cdf_attrs.yaml")
+    # Loading in test data
+    cdf_manager.load_variable_attributes("imap_test_variable.yaml")
 
     # All variables required to have:
     expected_attributes = [
+        "CATDESC",
+        "DEPEND_0",
         "DISPLAY_TYPE",
+        "FIELDNAM",
         "FILLVAL",
         "FORMAT",
+        "LABLAXIS",
+        "UNITS",
         "VALIDMIN",
         "VALIDMAX",
         "VAR_TYPE",
     ]
 
-    for variable_attrs in cdf_manager.variable_attributes.values():
-        for attr in expected_attributes:
-            assert attr in variable_attrs.keys()
+    # Assuring all required attributes are loaded in
+    for attr_name in cdf_manager.variable_attribute_schema["attribute_key"]:
+        attribute = cdf_manager.variable_attribute_schema["attribute_key"][attr_name]
+        if attribute["required"] is True:
+            for exp_attr in expected_attributes:
+                assert (
+                    exp_attr
+                    in cdf_manager.variable_attribute_schema["attribute_key"].keys()
```

Comment:
```suggestion
                    in cdf_manager.variable_attribute_schema["attribute_key"]
```

You don't need `.keys()`. I know this comes from my test but it's still true 😆 Greg always corrects me on this and it might have finally stuck now!